tests/readmes/test_erc20_enterprise.py

```diff
@@ -0,0 +1,273 @@
+        ),
+    )
+
+    erc20_enterprise_token.buy_from_dispenser_and_order(
+        order_params=order_params,
+        dispenser_address=dispenser.address,
+        from_wallet=alice_wallet,
+    )
+    increased_balance = OCEAN_token.balanceOf(bob_wallet.address)
+    assert initial_bob_balance < increased_balance
+
+
+def test_erc20_enterprise_flow_with_fre():
+    config = ExampleConfig.get_config()
+    ocean = Ocean(config)
+
+    # Create Alice's wallet
+    alice_private_key = os.getenv("TEST_PRIVATE_KEY1")
+    alice_wallet = Wallet(
+        ocean.web3,
+        alice_private_key,
+        config.block_confirmations,
+        config.transaction_timeout,
+    )
+    assert alice_wallet.address
+
+    # Mint OCEAN
+    mint_fake_OCEAN(config)
+    assert alice_wallet.web3.eth.get_balance(alice_wallet.address) > 0, "need ETH"
+
+    # Publish an NFT token
+    nft_token = ocean.create_nft_token("NFTToken1", "NFT1", alice_wallet)
+    token_address = nft_token.address
+    assert token_address
+
+    # Prepare data for ERC20 Enterprise token
+    cap = ocean.to_wei(200)
+    erc20_data = CreateErc20Data(
+        template_index=2,  # default value
+        strings=["ERC20DT1", "ERC20DT1Symbol"],  # name & symbol for ERC20 token
+        addresses=[
+            alice_wallet.address,  # minter address
+            alice_wallet.address,  # fee manager for this ERC20 token
+            alice_wallet.address,  # publishing Market Address
+            ZERO_ADDRESS,  # publishing Market Fee Token
+        ],
+        uints=[cap, 0],
+        bytess=[b""],
+    )
+    erc20_enterprise_token = nft_token.create_enterprise_datatoken(
+        erc20_data=erc20_data, from_wallet=alice_wallet
+    )
+
+    assert erc20_enterprise_token.address
+    assert erc20_enterprise_token.token_name() == "ERC20DT1"
+    assert erc20_enterprise_token.symbol() == "ERC20DT1Symbol"
+
+    bob_private_key = os.getenv("TEST_PRIVATE_KEY2")
+    bob_wallet = Wallet(
+        ocean.web3,
+        bob_private_key,
+        config.block_confirmations,
+        config.transaction_timeout,
+    )
+
+    # Verify that Bob has ganache ETH
+    assert ocean.web3.eth.get_balance(bob_wallet.address) > 0, "need ganache ETH"
+
+    # Bob buys 1 DT from the FRE and then startsOrder, while burning that DT
+    fixed_rate_exchange = ocean.fixed_rate_exchange
+    OCEAN_token = ocean.get_datatoken(ocean.OCEAN_address)
+
+    exchange_id = ocean.create_fixed_rate(
+        erc20_token=erc20_enterprise_token,
+        base_token=OCEAN_token,
+        amount=ocean.to_wei(25),
+        from_wallet=alice_wallet,
+    )
+
+    # Prepare data for order.
+    consume_fee_amount = ocean.to_wei(2)
+    provider_fee_address = alice_wallet.address
+    provider_fee_token = OCEAN_token.address
+    provider_fee_amount = 0
+    provider_data = json.dumps({"timeout": 0}, separators=(",", ":"))
+    valid_until = 1958133628  # 2032
+
+    message = ocean.web3.solidityKeccak(
```

Comment:
I'm seeing too much of this. Can we vreat some utilitary to create the provider mesage and signature based on the main things. The user shouldn't see stuff like keccak, signing etc. They should have a utility in Ocean class. Please create it and make sure we replace it in other readmes as well.